holoviews/streams.py

```diff
@@ -0,0 +1,166 @@
+"""
+The streams module defines the streams API that allows visualizations to
+generate and respond to events, originating either in Python on the
+server-side or in Javascript in the Jupyter notebook (client-side).
+"""
+
+import param
+import uuid
+from collections import OrderedDict
+
+
+class Stream(param.Parameterized):
+    """
+    A Stream is simply a parameterized object with parameters that
+    change over time in response to update events. Parameters are
+    updated via the update method.
+
+    Streams may have one or more subscribers, callables that are passed
+    the parameter dictionary when the trigger classmethod is called.
+    """
+
+    # Mapping from uuid to stream instance
+    registry = OrderedDict()
+
+    @classmethod
+    def trigger(cls, streams):
+        """
+        Given a list of streams, collect all the stream parameters into
+        a dictionary and pass it to the union set of subscribers.
+
+        Passing multiple streams at once to trigger can be useful when a
+        subscriber may be set multiple times across streams but only
+        needs to be called once.
+        """
+        # Union of stream values
+        items = [stream.value.items() for stream in streams]
+        union = dict(kv for kvs in items for kv in kvs)
+        # Currently building a simple set of subscribers
+        groups = [stream.subscribers + stream._hidden_subscribers for stream in streams]
+        subscribers = set(s for subscribers in groups for s in subscribers)
+        for subscriber in subscribers:
+            subscriber(union)
+
+    def update(self, trigger=True, **kwargs):
+        """
+        The update method updates the stream parameters in response to
+        some event.
+
+        If trigger is enabled, the trigger classmethod is invoked on
+        this particular Stream instance.
+        """
+        for param in self.params().values():
+            param.constant = False
+        self.set_param(**kwargs)
+        for param in self.params().values():
+            param.constant = True
```

Comment:
I'm a bit a alarmed by the suggestion to subclass from Stream for creating a parameterized object for use with paramnb -- the whole point of paramnb is to be able to display Jupyter-editable parameter widgets for parameters that are defined independently of any GUI or plotting system.  The underlying parameterized object sometimes is created just for the GUI, but in many cases is something that doesn't know or need to know anything about anything to do with GUIs or plotting.  Maybe in this particular case it is fine, because people using paramnb together with HoloViews streams are already doing things tied to plotting, but it does alarm me!